great_expectations/data_context/store/datasource_store.py

```diff
@@ -29,6 +37,14 @@ def __init__(
             store_name=store_name,
         )
 
+        # When using an InlineStoreBackend, we MUST use the same exact context (and not a copy).
+        # Equivalent contexts at different memory addresses will result in inconsistent persistence.
+        if (
+            isinstance(self._store_backend, InlineStoreBackend)
+            and store_backend is not None
+        ):
+            self._store_backend.data_context = store_backend.get("data_context")
```

Comment:
This was the gnarly bug that I spent 4-5 hours on today - the parent constructor called with `super()` creates a copy of the context instead of using the actual context. Changes to the config wouldn't persist consistently and we wouldn't fully delete values.

I don't love an explicit conditional to check for the type but I found this to be the cleanest way to approach it without modifying our existing Store instantiation pattern (which works for everything else).